build/generate-emoji.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Copyright 2015 Kenneth Shaw
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+// +build ignore
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"go/format"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+const (
+	gemojiURL = "https://raw.githubusercontent.com/github/gemoji/master/db/emoji.json"
+)
+
+var (
+	flagOut = flag.String("o", "modules/emoji/emoji_data.go", "out")
+)
+
+// Gemoji is a set of emoji data.
+type Gemoji []Emoji
+
+// Emoji represents a single emoji and associated data.
+type Emoji struct {
+	Emoji          string   `json:"emoji"`
+	Description    string   `json:"description"`
+	Category       string   `json:"category"`
+	Aliases        []string `json:"aliases"`
+	Tags           []string `json:"tags"`
+	UnicodeVersion string   `json:"unicode_version"`
+	IOSVersion     string   `json:"ios_version"`
+	SkinTones      bool     `json:"skin_tones"`
+}
+
+func main() {
+	var err error
+
+	flag.Parse()
+
+	// generate data
+	buf, err := generate()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// write
+	err = ioutil.WriteFile(*flagOut, buf, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+var replacer = strings.NewReplacer(
+	"main.Gemoji", "Gemoji",
+	"main.Emoji", "\n",
+	"}}", "},\n}",
+	", Description:", ", ",
+	", Category:", ", ",
+	", Aliases:", ", ",
+	", Tags:", ", ",
+	", UnicodeVersion:", ", ",
+	", IOSVersion:", ", ",
+	", SkinTones:", ", ",
+)
+
+var emojiRE = regexp.MustCompile(`\{Emoji:"([^"]*)"`)
+
+func generate() ([]byte, error) {
+	var err error
+
+	// load gemoji data
+	res, err := http.Get(gemojiURL)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	// read all
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// unmarshal
+	var data Gemoji
+	err = json.Unmarshal(body, &data)
+	if err != nil {
+		return nil, err
+	}
+
+	// add header
+	str := replacer.Replace(fmt.Sprintf(hdr, gemojiURL, data))
+
+	// change the format of the unicode string
+	str = emojiRE.ReplaceAllStringFunc(str, func(s string) string {
+		var err error
+		s, err = strconv.Unquote(s[len("{Emoji:"):])
+		if err != nil {
+			panic(err)
+		}
+		return "{" + strconv.QuoteToASCII(s)
+	})
+
+	// write a JSON file to use with tribute
+	file, _ := json.MarshalIndent(data, "", " ")
+	_ = ioutil.WriteFile("public/emoji/emoji.json", file, 0644)
+
+	// format
+	return format.Source([]byte(str))
+}
+
+const hdr = `
+package emoji
```

Comment:
Isnt the copiright header first and the package definde?

Reply:
Fixed